service/src/main/java/bio/terra/workspace/service/resource/controlled/cloud/azure/vm/AssignActionManagedIdentitiesVmStep.java

```diff
@@ -0,0 +1,107 @@
+package bio.terra.workspace.service.resource.controlled.cloud.azure.vm;
+
+import bio.terra.stairway.FlightContext;
+import bio.terra.stairway.Step;
+import bio.terra.stairway.StepResult;
+import bio.terra.stairway.exception.RetryException;
+import bio.terra.workspace.app.configuration.external.AzureConfiguration;
+import bio.terra.workspace.common.utils.Rethrow;
+import bio.terra.workspace.service.crl.CrlService;
+import bio.terra.workspace.service.iam.AuthenticatedUserRequest;
+import bio.terra.workspace.service.iam.SamService;
+import bio.terra.workspace.service.iam.model.SamConstants;
+import bio.terra.workspace.service.workspace.flight.WorkspaceFlightMapKeys;
+import bio.terra.workspace.service.workspace.model.AzureCloudContext;
+import com.azure.resourcemanager.compute.ComputeManager;
+import com.azure.resourcemanager.msi.MsiManager;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Queries Sam for any action managed identities for the calling user, and assigns them to the VM.
+ */
+public class AssignActionManagedIdentitiesVmStep implements Step {
+  private static final Logger logger =
+      LoggerFactory.getLogger(AssignActionManagedIdentitiesVmStep.class);
+
+  private final AzureConfiguration azureConfig;
+  private final CrlService crlService;
+  private final SamService samService;
+  private final AuthenticatedUserRequest userRequest;
+  private final ControlledAzureVmResource resource;
+
+  public AssignActionManagedIdentitiesVmStep(
+      AzureConfiguration azureConfig,
+      CrlService crlService,
+      SamService samService,
+      AuthenticatedUserRequest userRequest,
+      ControlledAzureVmResource resource) {
+    this.azureConfig = azureConfig;
+    this.crlService = crlService;
+    this.samService = samService;
+    this.userRequest = userRequest;
+    this.resource = resource;
+  }
+
+  @Override
+  public StepResult doStep(FlightContext context) throws InterruptedException, RetryException {
+    final AzureCloudContext azureCloudContext =
+        context
+            .getWorkingMap()
+            .get(
+                WorkspaceFlightMapKeys.ControlledResourceKeys.AZURE_CLOUD_CONTEXT,
+                AzureCloudContext.class);
+
+    ComputeManager computeManager = crlService.getComputeManager(azureCloudContext, azureConfig);
+    MsiManager msiManager = crlService.getMsiManager(azureCloudContext, azureConfig);
+
+    // Query Sam for azure_managed_identity resources the user has access to
```

Comment:
TODO: potentially remove this, align on a convention that the AAMI resource is named with the billing profile ID